### ISSUE TITLE
Bug with commit in custom branch

### ISSUE THREAD

**Issue by yuriydzobak** [fluxcd/image-automation-controller#161]
Hi
Does anybody face an issue with ImageUpdateAutomation when you use `update.path`
I’ve enabled `update.path`  and contoller has create a custom branch with changes in custom path, but sometimes, the contoller pushed commit that modify only tag under `update.path` but sometimes it pushed commit that modify `tag` and remove all files that wasn’t included in `update.path`
for example, first and second  commit modify files in `env/cluster` and doesn’t touch any other files in `/*`  but sometimes it modify file in `env/cluster`  and remove other files from `/*`

<img width="1447" alt="image" src="https://user-images.githubusercontent.com/70581130/117124626-41c98d80-ada1-11eb-9120-3eb5acaf5541.png">

Also, it happed in different flux contollers
I have deployed flux in two clusters and that issue show up

```
apiVersion: image.toolkit.fluxcd.io/v1alpha2
kind: ImageUpdateAutomation
metadata:
  name: image-update-automation
spec:
  interval: 5m
  sourceRef:
    kind: GitRepository
    name: github
  git:
    checkout:
      ref:
        branch: master
    push:
      branch: flux-image-update-cluster1
    commit:
      author:
        name: fluxcdbot
        email: fluxcdbot@users.noreply.github.com
      messageTemplate: |
        [ci-skip] Automated image update

        Automation name: {{ .AutomationObject }}

        Files:
        {{ range $filename, $_ := .Updated.Files -}}
        - {{ $filename }}
        {{ end -}}

        Objects:
        {{ range $resource, $_ := .Updated.Objects -}}
        - {{ $resource.Kind }} {{ $resource.Name }}
        {{ end -}}

        Images:
        {{ range .Updated.Images -}}
        - {{.}}
        {{ end -}}
  update:
    strategy: Setters
    path: "env/cluster1"
```

I upgrade the controller to latest version(v0.9.0), but it didn't help

**Comment by squaremo** [comment]
@yuriydzobak Thanks for reporting this!

 - Is the git repository available for me to look at?
 - Do the commit messages for the problem commits look different to the others, or otherwise weird?
 - Did it report any errors in the controller logs?

**Comment by yuriydzobak** [comment]
> @yuriydzobak Thanks for reporting this!
> 
> * Is the git repository available for me to look at?
- nope
> * Do the commit messages for the problem commits look different to the others, or otherwise weird?
Commit message sometimes is empty
<img width="632" alt="image" src="https://user-images.githubusercontent.com/70581130/117178437-b2d96700-adda-11eb-99c7-9203966e4d8d.png">

sometimes it has update the same file and the second commit delete all files =(
example
```
Automation name: foo/image-update-automation-foo

Files:
- application-services/services/foo-bar-poller/image-tag.yaml
Objects:
- HelmRelease foo-bar-poller
Images:
- ecr.foo.info/foo-bar-poller:release.00cdc5549-20210505-0848
@fluxcdbot
fluxcdbot committed 8 hours ago

[ci-skip] Automated image update  …
Automation name: foo/image-update-automation-foo

Files:
- application-services/services/foo-bar-poller/image-tag.yaml
Objects:
- HelmRelease foo-bar-poller
Images:
- ecr.foo.com/foo-bar-poller:release.00cdc5549-20210505-0913
```

> * Did it report any errors in the controller logs?

controller logs don't have any entries, just information about that pushed commit to the branch

**Comment by squaremo** [comment]
Without knowing enough to reproduce the problem directly, I'm examining the code for any places which could result in this kind of commit.

The procedure `commitChangedManifests` was my first suspect. However it will already return an error if it's given a working tree with removed files, since it `lstat`s every file mentioned in the changeset, and if a file were removed that would fail.

With that in mind: @yuriydzobak is there definitely a commit that removes files, or are you seeing the diff between the automation branch `flux-image-update-cluster1` and the pull request target `master`? In which case, we can try to figure out whether this is a flaw in the implementation or in the design of the API.

**Comment by yuriydzobak** [comment]
Only this error displays
```
{"level":"error","ts":"2021-05-27T12:04:14.926Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"Reconciler error","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","error":"file not found"}
{"level":"error","ts":"2021-05-27T12:04:34.709Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"Reconciler error","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","error":"file not found"}
{"level":"error","ts":"2021-05-27T12:04:53.061Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"Reconciler error","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","error":"file not found"}
{"level":"error","ts":"2021-05-27T12:05:13.120Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"Reconciler error","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","error":"file not found"}
{"level":"error","ts":"2021-05-27T12:05:32.153Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"Reconciler error","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","error":"file not found"}
{"level":"error","ts":"2021-05-27T12:05:53.707Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"Reconciler error","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","error":"file not found"}
{"level":"error","ts":"2021-05-27T12:06:12.846Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"Reconciler error","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","error":"file not found"}
{"level":"error","ts":"2021-05-27T12:06:31.006Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"Reconciler error","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","error":"file not found"}
{"level":"error","ts":"2021-05-27T12:06:53.414Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"Reconciler error","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","error":"file not found"}
```

**Comment by squaremo** [comment]
> Only this error displays

Has that error been logged all the time you've seen these commits?

**Comment by yuriydzobak** [comment]
So, when i recreated POD, the issue was gone but when commit has been made with delete of all files, this error is showing

**Comment by yuriydzobak** [comment]
I've enabled debug mode, we'll see if i can provide addition information

**Comment by yuriydzobak** [comment]
debug mode, when controller has pushed commit with remove files, nothing special
```
{"level":"debug","ts":"2021-05-27T18:35:05.740Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"ran updates to working dir","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","working":"/tmp/application-github-repo-cm-application551644299"}
{"level":"debug","ts":"2021-05-27T18:35:08.416Z","logger":"controller-runtime.manager.events","msg":"Normal","object":{"kind":"ImageUpdateAutomation","namespace":"application","name":"image-update-automation-application","uid":"8fa67f54-604c-4039-acb3-df963456a013","apiVersion":"image.toolkit.fluxcd.io/v1alpha2","resourceVersion":"97662984"},"reason":"info","message":"no updates made"}
{"level":"debug","ts":"2021-05-27T18:35:08.433Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"no changes made in working directory; no commit","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application"}
{"level":"debug","ts":"2021-05-27T18:40:09.603Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"found git repository","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","gitrepository":{"namespace":"application","name":"github-repo-cm-application"}}
{"level":"debug","ts":"2021-05-27T18:40:24.483Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"cloned git repository","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","gitrepository":{"namespace":"application","name":"github-repo-cm-application"},"ref":{"branch":"master"},"working":"/tmp/application-github-repo-cm-application235137488"}
{"level":"debug","ts":"2021-05-27T18:40:24.496Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"ran updates to working dir","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","working":"/tmp/application-github-repo-cm-application235137488"}
{"level":"debug","ts":"2021-05-27T18:40:26.594Z","logger":"controller-runtime.manager.events","msg":"Normal","object":{"kind":"ImageUpdateAutomation","namespace":"application","name":"image-update-automation-application","uid":"8fa67f54-604c-4039-acb3-df963456a013","apiVersion":"image.toolkit.fluxcd.io/v1alpha2","resourceVersion":"97666836"},"reason":"info","message":"no updates made"}
{"level":"debug","ts":"2021-05-27T18:40:26.634Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"no changes made in working directory; no commit","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application"}
{"level":"debug","ts":"2021-05-27T18:45:27.692Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"found git repository","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","gitrepository":{"namespace":"application","name":"github-repo-cm-application"}}
{"level":"debug","ts":"2021-05-27T18:45:44.716Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"cloned git repository","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","gitrepository":{"namespace":"application","name":"github-repo-cm-application"},"ref":{"branch":"master"},"working":"/tmp/application-github-repo-cm-application041476857"}
{"level":"debug","ts":"2021-05-27T18:45:44.718Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"ran updates to working dir","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","working":"/tmp/application-github-repo-cm-application041476857"}
{"level":"debug","ts":"2021-05-27T18:45:50.654Z","logger":"controller-runtime.manager.events","msg":"Normal","object":{"kind":"ImageUpdateAutomation","namespace":"application","name":"image-update-automation-application","uid":"8fa67f54-604c-4039-acb3-df963456a013","apiVersion":"image.toolkit.fluxcd.io/v1alpha2","resourceVersion":"97670576"},"reason":"info","message":"committed and pushed change eb6baf049005f7dd55f7c097b0d5fe6be51226fa to flux-image-update-application"}
{"level":"info","ts":"2021-05-27T18:45:50.668Z","logger":"controller-runtime.manager.controller.imageupdateautomation","msg":"pushed commit to origin","reconciler group":"image.toolkit.fluxcd.io","reconciler kind":"ImageUpdateAutomation","name":"image-update-automation-application","namespace":"application","revision":"eb6baf049005f7dd55f7c097b0d5fe6be51226fa","branch":"flux-image-update-application"}
```

**Comment by therapy-lf** [comment]
This bug makes controller unusable for us. Any suggestion/thoughts on how we can go further on debugging this problem?

**Comment by squaremo** [comment]
> Any suggestion/thoughts on how we can go further on debugging this problem?

@therapy-lf At present I cannot reproduce the problem as described. If you can create a minimal (or at least small) example of a git repository and API objects that lead to the problem described, that will help a great deal.

**Comment by yuriydzobak** [comment]
Hi,
I created the repo and error was reproduced
invite I've sent you https://github.com/kolikons/flux

**Comment by squaremo** [comment]
With a bit of experimentation and help from @yuriydzobak I have been able to reproduce the problem with image-automation-controller v0.9.0:

 1. set up a git repo with an image update automation using both a `.path` and a `.push.branch`
 2. provoke an automated commit to the push branch
 3. make a change outside the path and commit that to the main (checkout) branch
 4. provoke another automated commit

This results in a broken local git repo in which files are missing, in 4.). Depending on the change made in 3.), this will either result in no commit when there should have been one (because no files under .path to change), or a commit which removes a load of files.

Looking at the repo as checked out by the controller, it seems like the cause of the brokenness is some combination of the shallow clone, and code in `switchBranch` that tries to check out the push branch. The good news is that in v0.11.0, the controller always uses libgit2 to clone, so it's never a shallow clone (I also simplified the code in `switchBranch`). I cannot get it to break with the above steps, when running the code in v0.11.0 (or from main branch).
